source/ways-of-working/standards/db-schema-dev-workflow.html.md.erb

```diff
@@ -0,0 +1,75 @@
+Essentially, database migrations must be done in such a way that they are done independently of application deployments. The underlying principle to obey here is:
+Every change you make must be backward compatible with the rest of the system
+
+For the add non-nullable column scenario these are the backward compatible steps to perform the change:
+
+1. Add nullable column to database – System keeps adding rows, nulls are fine, reads ignore the null
+2. Code change to write correct value to new rows, and handle reading unexpected nulls – Database doesn’t change, now we have some null rows and some rows with data
+3. Run data migration to fill the other columns – This might be a script, or a bit of code in the application, either way your app doesn’t care about any row, it handles data and nulls just fine
+4. Add the non-null constraint – The database now has no nulls and your new code is writing the correct data.
+5. Remove the code that handles the null case – it won’t happen anymore.
+
+![Zero downtime](/images/zero-downtime-db-migrations.png)
+
+## Strategies for refactoring databases
+More strategies for breaking large database schema changes down into smaller, less-risky steps can be found in Refactoring Databases: Evolutionary Database Design (Pramod J. Sadalage, Scott W. Ambler, 2006).
+
+Chapter 5 of that book gives an introduction to those strategies and 'lessons learned' for doing applying evolutionary database design.
+
+- Smaller changes are easier to apply.
+- Uniquely identify individual refactorings.
+- Implement a large change by many small ones.
+- Have a database configuration table.
+- Prefer triggers over views or batch synchronization.
+- Choose a sufficient deprecation period.
+- Simplify your database change control board (CCB) strategy.
+- Simplify negotiations with other teams.
+- Encapsulate database access.
+- Be able to easily set up a database environment.
+- Do not duplicate SQL.
+- Put database assets under change control.
+- Beware of politics.
+
+The last point probably requires an explanation:
+
+> Introducing evolutionary database techniques, and in particular database refactoring, is likely to be a major change within your organization. As Manns and Rising (2005) point out in Fearless Change, politics is likely to rear its ugly head as you try to make these improvements to the way that you work. Although many IT professionals prefer to ignore politics, naively believing that they are above it, they do so at their peril. The techniques that we describe in this book work; we know this because we have done it. We also know that many traditional data professionals are threatened by these techniques, and rightfully so, because these techniques force them to make significant changes to the way that they will work in the future. You will likely need to “play the political game” if you want to adopt database refactoring within your organization.
```

Comment:
it's already a part of the organisation do we need to mention this, it's a fair bit of text